### PR TITLE
libbeat/reader/auditd: add multi-record coalescing mode

### LINKED ISSUE #52556: [auditd] Add multi-record coalescing to filestream parser
Extends the Phase 1 auditd filestream parser (`libbeat/reader/auditd`,
landed in #50791) with multi-record coalescing so the `auditd`
integration produces compound events compatible with the
`auditd_manager` field schema.

Parent: https://github.com/elastic/integrations/issues/19080

## Context

131 detection rules target `auditd_manager` exclusively; zero target
`auditd`. 101 of those rules require compound-event fields
(`process.args`, `process.title`, `file.path`, `auditd.summary.*`)
that only exist when multiple audit records are assembled into a single
document. No per-record approach can reach these rules.

The Phase 1 parser calls `auparse.ParseLogLine()` per line and emits
one `reader.Message` per record. This issue adds a second stage that
groups records by sequence number and coalesces them into compound
events using go-libaudit's `Reassembler` and `aucoalesce`.

## Work

- Add `Reassembler` integration and a ~channel-based~ `Next()` adapter
  bridging the asynchronous reassembler callback to the synchronous
  `reader.Reader` interface.
- Call `aucoalesce.CoalesceMessages()` on completed groups.
- Convert the resulting `aucoalesce.Event` into a `reader.Message`
  with compound fields in the `auditd.data.*` namespace.
- Handle shutdown, context cancellation, and partial flushes on file
  rotation (`Reassembler.Close()`).
- Preserve the `node=` log header value as `auditd.data.node`.
- Add a `parser_mode` config setting (`none | parse | coalesce`,
  default `none`) replacing the existing `use_auditd_parser` boolean.
  Include a migration note in the changelog.

## References

- Phase 1 parser: #50791
- Phase 1 integration toggle: elastic/integrations#19071
- go-libaudit packages: [Reassembler](https://pkg.go.dev/github.com/elastic/go-libaudit/v2#Reassembler), [aucoalesce](https://pkg.go.dev/github.com/elastic/go-libaudit/v2/aucoalesce)

### PR DESCRIPTION
**Pull request by efd6**
## Proposed commit message
```
libbeat/reader/auditd: add multi-record coalescing mode

Add a "coalesce" parser mode that groups related audit records by
sequence number using go-libaudit's Reassembler and produces compound
events via aucoalesce.CoalesceMessages. The output uses the
auditd.data.* namespace (matching the auditd_manager integration) and
populates ECS root fields (process.*, user.*, event.*, file.*).

The parser now supports three modes controlled by a "mode" config
field: "parse" (default, per-line, the existing behaviour), "coalesce"
(multi-record grouping), and "none" (pass-through). The design is
synchronous, using a slice queue flushed by the existing DeadlineSetter
timeout mechanism rather than channels or background goroutines.
```

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works. Where relevant, I have used the [`stresstest.sh`](https://github.com/elastic/beats/blob/main/script/stresstest.sh) script to run them under stress conditions and race detector to verify their stability.
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent-changelog-tool/blob/main/docs/usage.md).

## Related issues

- Closes #52556

### CODE REVIEW

File: docs/reference/filebeat/filebeat-input-filestream.md [hunk 1]
```diff
@@ -1027,13 +1027,20 @@ This parser is only supported on Linux. On other platforms, configuring it retur
 
 The supported configuration options are:
 
+**`mode`**
+:   {applies_to}`stack: ga 9.5.3+` (Optional) Controls the parser behaviour. Valid values:
```
Comment (vishaangelova) [inline docs/reference/filebeat/filebeat-input-filestream.md]: Let’s move the `applies_to` as it applies to the whole “term” ([guideline](https://www.elastic.co/docs/contribute-docs/how-to/cumulative-docs/badge-placement#definition-list-item-full)) + adjust to American English ([guideline](https://www.elastic.co/docs/contribute-docs/style-guide#grammar-and-spelling)):

```suggestion
**`mode`** {applies_to}`stack: ga 9.5.3+`
:   (Optional) Controls the parser behavior. Valid values:
```

File: docs/reference/filebeat/filebeat-input-filestream.md [hunk 2]
```diff
@@ -1027,13 +1027,20 @@ This parser is only supported on Linux. On other platforms, configuring it retur
...
+
+    - `parse` (default): Each audit log line is parsed individually. Fields are added under `auditd.log.*`. This preserves one output event per input line.
+    - `coalesce`: Related audit records sharing the same sequence number are grouped into a single compound event using the same logic as `auditd_manager`. Fields are added under `auditd.data.*`, `auditd.summary.*`, and ECS root fields (`process.*`, `user.*`, `file.*`, etc.). Incomplete groups are flushed after a 2-second timeout.
```
Comment (vishaangelova) [inline docs/reference/filebeat/filebeat-input-filestream.md]: We try to avoid Latin abbreviations where possible ([guideline](https://www.elastic.co/docs/contribute-docs/style-guide/accessibility#write-for-an-international-audience)), so maybe:

```suggestion
    - `coalesce`: Related audit records sharing the same sequence number are grouped into a single compound event using the same logic as `auditd_manager`. Fields are added under `auditd.data.*`, `auditd.summary.*`, and ECS root fields (`process.*`, `user.*`, `file.*`, and so on). Incomplete groups are flushed after a 2-second timeout.
```

**Review by vishaangelova — APPROVED** [review]
LGTM, but let’s move the `applies_to` tag

File: libbeat/reader/auditd/coalesce.go [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+		if nodeVal != "" {
+			p.node = nodeVal
+		}
+
+		auditMsg, parseErr := auparse.ParseLogLine(line)
+		if parseErr != nil {
+			if p.cfg.LogErrors {
+				p.logger.Debugw("skipping unparseable line", "error", parseErr)
+			}
+			continue
```
Comment (muskan-agarwal26) [inline libbeat/reader/auditd/coalesce.go:172]: In coalesce mode, unparseable lines and CoalesceMessages failures are dropped (debug log only). parse mode still emits those lines. For a security log source, can we emit a fallback event with error.message (and the raw line/group) instead of silently losing data? If dropping is intentional, that should be documented.

File: libbeat/reader/auditd/coalesce.go [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+		}
+		p.reasm.PushMessage(auditMsg)
+	}
+}
+
+// distributeBytesRead spreads accumulated bytesRead evenly across all pending
+// events so that each has a non-zero Bytes value. The remainder goes to the
+// first event (the one about to be returned).
+func (p *coalescingParser) distributeBytesRead() {
```
Comment (muskan-agarwal26) [inline libbeat/reader/auditd/coalesce.go:190]: If two groups are in flight and the first one completes, bytesRead (including the second group’s lines) is all assigned to the completed event. The remaining group can later flush on timeout with Bytes=0 and get dropped by filestream; which this comment says we are trying to avoid. `TestCoalescingInterleavedBytesAccounting` only covers both groups flushing together. Could we add a test that completes seq A, then times out seq B, and assert B still has non-zero Bytes?

File: libbeat/reader/auditd/coalesce.go (hunk 3 above)
Comment (efd6) [reply libbeat/reader/auditd/coalesce.go:172]: This is intentional and reflects the statements in the issue and the RFC, so I don't think it needs a comment.

File: libbeat/reader/auditd/coalesce.go (hunk 4 above)
Comment (efd6) [reply libbeat/reader/auditd/coalesce.go:190]: Agree, will fix.

**Review by muskan-agarwal26 — APPROVED** [review]
Thanks

File: docs/reference/filebeat/filebeat-input-filestream.md
Comment (kcreddy) [inline docs/reference/filebeat/filebeat-input-filestream.md]: This needs correction, as its now pass-through on non-linux as per `auditd_other.go`

Comment (kcreddy) [inline docs/reference/filebeat/filebeat-input-filestream.md]: Could update this line as the new `coalesce` mode adds the fields under `auditd.data.*`

**Review by kcreddy — APPROVED** [review]
Nits on documentation. LGTM otherwise
